src/rabbitalerter.js

```diff
@@ -0,0 +1,208 @@
+/**
+ * An alerting system for message queues.
+ * @module rabbitalerter
+ */
+const assert = require('assert');
+const taskcluster = require('taskcluster-client');
+
+/**
+ * An interface used to generate various TaskCluster Notifications.
+ *
+ * @class RabbitAlerter
+ */
+class RabbitAlerter {
+  /**
+   * @param {number} config.alerter.messageCountTolerance       - When a queue holds more messages
+   *                                                              than this tolerance, an alert will be created.
+   * @param {number} config.alerter.messagePublishRateTolerance - When the number of messages published per second
```

Comment:
I suspect these are the wrong metrics... to alert on..

I think we want to alert when:
 A) **the consumer can't keep up with the number of messages published**.
 B) A non-empty queue haven't had messages consumed for a significant amount of time, say 8 minutes.

Ignoring (A) will eventually cause a full system failure (out of disk/memory).
Where as (B) is just an indicator that something crashed, for example a low through-put queue whose consumer crashed.

A fixed upper bound solves the problem for (A), until you have a queue with a really high through-put, then it might spike the fixed limit and trigger an alert. Further more the higher the fixed limit is the worse this detection is for case (B).

It's not obvious to me how to do this... Because we're only observing queue stats for points in time... We can't tell how many messages was consumed since last we looked at the stats.
Technically we have: https://gist.github.com/jonasfj/925b8652b015ff857c629cef6106fac7
(that's from a high through-out queue).

My initial thinking was that if observed that queue at every 1 min interval, if for 30 consecutive intervals there never is zero messages in the queue, then it's not being consumed.
But my strategy certainly has other downsides, such as it might not work for high-through-put queues.

Maybe the strategy is to say:
Let x[i] be the number of messages in the queue at interval i, and let y be the median of x[j], x[j+1], ... x[j+k],
if y < min(x[j+k], x[j+k-1], x[j+k-c]) then we raise an alert, where j + k is the current interval and c < k.

Essentially saying if for some c number of intervals, the message count haven't been lower than the median message count for the last k number of intervals, then we raise an alert. For this k might have to be large enough to quantify over the past week, and c might be low enough to only consider intervals from the past 30 minutes or so...

Maybe average is easier than median... who knows... Maybe there is a better option here... This was just a random thought. But this is the hard part, how do we detect that message count is trending upwards overtime, if we only observe the message count at certain snapshots.

Whatever strategy we come up with, two alerts are probably safe to do:
 1) Queue has reach some obscenely high number of messages (say 1M)
 2) Queue haven't seen any consumption in say 25 minutes or so. 

While those are definitely sane, (1) won't catch crashed consumers soon enough, and never catch low-through-put queues. (2) won't do anything high-through-put queues that consumers can't keep up with.

------

Furthermore, I suspect we want to alert people if a queue is non-empty and nothing has been consumed for a significant amount of time, say 8 minutes.

Small queues with only a few messages that never gets consumed or deleted aren't going to cause system failures.
But they are indicators that something broke somewhere, so it's nice of us to alert people about that.